File: src/foundry/schemas/agent.py
"""Coding-agent job contracts.

These are provider-agnostic: every ``CodingAgentProvider`` (Cursor, Claude Code,
OpenAI agent, manual) accepts the same ``CodingAgentJobInput`` and reports the
same ``CodingAgentJob`` shape. Secrets must never travel in these objects.
"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from .common import AgentJobStatus
from .plan import DeliveryPlan


class JobConstraints(BaseModel):
    model_config = ConfigDict(extra="forbid")

    do_not_modify: list[str] = Field(default_factory=list)
    required_tests: list[str] = Field(default_factory=list)
    max_files_changed: int = Field(default=12, ge=1)
    allow_new_dependencies: bool = False


class CodingAgentJobInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    repo: str
    base_branch: str = "main"
    branch_name: str
    ticket_url: str
    delivery_plan: DeliveryPlan
    agent_instructions: str
    constraints: JobConstraints = Field(default_factory=JobConstraints)


class CodingAgentJob(BaseModel):
    """Handle returned when a job is created with a provider."""

    model_config = ConfigDict(extra="forbid")

    job_id: str
    provider: str
    status: AgentJobStatus = AgentJobStatus.CREATED


class CodingAgentJobStatus(BaseModel):
    model_config = ConfigDict(extra="forbid")

    job_id: str
    provider: str
    status: AgentJobStatus
    branch: str | None = None
    pr_url: str | None = None
    error: str | None = None

File: dotnet/src/Foundry/Schemas/Agent.cs
// Coding-agent job contracts.
//
// These are provider-agnostic: every ICodingAgentProvider (Cursor, Claude Code,
// OpenAI agent, manual) accepts the same CodingAgentJobInput and reports the
// same CodingAgentJob shape. Secrets must never travel in these objects.

namespace Foundry.Schemas;

public sealed record JobConstraints
{
    private readonly int _maxFilesChanged = 12;

    public IReadOnlyList<string> DoNotModify { get; init; } = Array.Empty<string>();

    public IReadOnlyList<string> RequiredTests { get; init; } = Array.Empty<string>();

    public int MaxFilesChanged
    {
        get => _maxFilesChanged;
        init => _maxFilesChanged = Guard.Min(value, 1, nameof(MaxFilesChanged));
    }

    public bool AllowNewDependencies { get; init; }
}

public sealed record CodingAgentJobInput
{
    public required string RunId { get; init; }

    public required string Repo { get; init; }

    public string BaseBranch { get; init; } = "main";

    public required string BranchName { get; init; }

    public required string TicketUrl { get; init; }

    public required DeliveryPlan DeliveryPlan { get; init; }

    public required string AgentInstructions { get; init; }

    public JobConstraints Constraints { get; init; } = new();

    /// <summary>
    /// Identifier of the source issue in the tracker (e.g. Linear), so a provider
    /// that delegates via the tracker (Cursor-via-Linear) can address it.
    /// </summary>
    public string? TrackerIssueId { get; init; }
}

/// <summary>Handle returned when a job is created with a provider.</summary>
public sealed record CodingAgentJob
{
    public required string JobId { get; init; }

    public required string Provider { get; init; }

    public AgentJobStatus Status { get; init; } = AgentJobStatus.Created;
}

public sealed record CodingAgentJobStatus
{
    public required string JobId { get; init; }

    public required string Provider { get; init; }

    public required AgentJobStatus Status { get; init; }

    public string? Branch { get; init; }

    public string? PrUrl { get; init; }

    public string? Error { get; init; }

    /// <summary>
    /// Spend reported by the provider for this job, where the provider exposes
    /// usage (e.g. the Cursor Cloud Agents API). Null = unknown, not zero.
    /// </summary>
    public double? CostUsd { get; init; }
}
